File: Gateway/parser.js


//parser to get data from header
function header_parser(buffer){

  var header_data = [];
  //parse sender id from the header
  header_data[0] = buffer.readUInt32LE(0);
  //parse app id from header
  header_data[1] = buffer.readUInt8(4);
  var val_5 = buffer.readUInt8(5) & 0xFF;
  //parse message_type from header
  header_data[2] = ( val_5 & 0b11100000)>>5;
  //parse ttl value from header
  header_data[3] = (val_5 & 0b00011100)>>2;
  //parse more_frame bit from header
  header_data[4] = (val_5 & 0b00000010)>>1;
  //parse seq# from header
  header_data[5] =  (val_5 & 0b00000001)<<8 | buffer.readUInt8(6) & 0xFF;

  return(header_data);
}


//generating message header for LoRa
function  generate_header(header_data){
  var header_buff =  new Buffer(7);
  //add sender id to header
  for(var i=3; i>=0; i--)    // start with lowest byte of number
  {
    header_buff[i] = header_data[0] & 0xFF;  // or: = byte( number);
    header_data[0] >>= 8;            // get next byte into position
  }
  //add app id to headerheader_arr
  header_buff[4] = header_data[1] & 0xFF;
  //set last two bytes of header
  header_buff[5] = (header_data[2] << 5) | (header_data[3] << 2) | (header_data[4] << 1) | (header_data[5] >> 8);
  header_buff[6] = header_data[5] & 0xFF;
  return header_buff;
  }

File: LoRa_End_Node/end_node_mcu/parser.cpp

#include "parser.h"



//-------------generating message header for LoRa--------------

char*  generate_header(struct Header header, char header_arr[]){

  //add sender id to header
  for(int i=3; i>=0; i--)    // start with lowest byte of number
  {
    header_arr[i] = header.sender & 0xFF;  //convert to bytes
    header.sender >>= 8;            // get next byte into position
  }

  //add app id to header
  header_arr[4] = header.app;
  //set last two bytes of header
  header_arr[5] = (header.type << 5) | (header.ttl << 2) | (header.mf << 1) | (header.seq >> 8);
  header_arr[6] = header.seq;
  //header_arr[7] = 7;
  return header_arr;
  }

  

//--------------parser to get data from header-----------------

struct Header header_parser(char* data, struct Header header){

//parse sender id from the header
 uint32_t sender=0;
 for(int i=3; i>=0; i--)    // start with lowest byte of number
  {
    sender += uint32_t(data[i]) << 8*(3-i);  //calculate value od 32bit data step by spep
  }
  header.sender=sender;

  //parse app id from header
  header.app=data[4];
  //parse message_type from header
  header.type=(data[5] & 0b11100000)>>5;
  //parse ttl value from header
  header.ttl=(data[5] & 0b00011100)>>2;
  //parse more_frame bit from header
  header.mf=(data[5] & 0b00000010)>>1;
  //parse seq# from header
  header.seq =  uint16_t(data[5] & 0b00000001)<<8 | data[6];

  return header;
}
